R/filter2.R

```diff
@@ -45,10 +45,21 @@ filter2 = function(x, filter, boundary = c("circular", "replicate")) {
          },
          ## pad with a given value
          linear = {
-           dx[1:2] = dx[1:2] + cf[1:2]
-           xpad = array(val, dx)
-           ## is there a better way of doing this?
-           x = do.call("[<-", c(quote(xpad), lapply(d, function(x) enquote(1:x)), quote(x)) )
+           x = imageData(x)
+           
+           dx[1:2] = dx[1:2] + df[1:2] - 1L
```

Comment:
Please note that my pull request does not extend on both sides. It only pads on the right bottom. The do.call is functionally equivalent between our branches (we both pad on the right bottom side).

The code to extract the filtered image is identical as well:
```R
  if ( boundary!="circular" ) {
    y = asub(y, list(1:d[1L], 1:d[2L]), 1:2)
  }
```

The only difference is the amount of padding we do. In your case you pad a different amount between boundary=linear `(cf = df %/%2)` and boundary=replicate `(df-1)`. I think the padding size should be the same, only the padding values should be different, depending on the method.

The padding should be the size of the filter kernel minus the center point, therefore:
```R
dx[1:2] = dx[1:2] + df[1:2] - 1L
```
for both linear and replicate boundary options.